File: android/app/src/main/java/com/tisza/esemenynaptar/EventListAdapter.java
package com.tisza.esemenynaptar;

import android.content.*;
import android.text.*;
import android.text.method.*;
import android.view.*;
import android.widget.*;
import androidx.lifecycle.Observer;
import androidx.lifecycle.*;
import com.tisza.esemenynaptar.database.*;

import java.util.*;

public class EventListAdapter extends BaseAdapter
{
	private LiveData<List<Event>> events;
	private TextView noEventsView = null;
	private Observer<List<Event>> observer = (list -> notifyDataSetChanged());

	public void setEvents(LiveData<List<Event>> newEvents)
	{
		if (events != null)
			events.removeObserver(observer);
		this.events = newEvents;
		events.observeForever(observer);
	}

	private List<Event> getCurrentEvents()
	{
		return events == null || events.getValue() == null ? Collections.emptyList() : events.getValue();
	}
	
	@Override
	public int getCount()
	{
		return Math.max(1, getCurrentEvents().size());
	}

	@Override
	public Object getItem(int position)
	{
		return null;
	}

	@Override
	public long getItemId(int position)
	{
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent)
	{
		LayoutInflater inflater = LayoutInflater.from(parent.getContext());

		if (getCurrentEvents().isEmpty())
		{
			if (noEventsView == null)
				noEventsView = (TextView)inflater.inflate(R.layout.no_events_view, parent, false);

			return noEventsView;
		}

		View view;
		ViewHolder viewHolder;
		if (convertView != null && convertView != noEventsView)
		{
			view = convertView;
			viewHolder = (ViewHolder)convertView.getTag();
		}
		else
		{
			view = inflater.inflate(R.layout.event_view, parent, false);
			viewHolder = new ViewHolder();
			viewHolder.categoryTextView = view.findViewById(R.id.event_category_text);
			viewHolder.iconView = view.findViewById(R.id.event_icon);
			viewHolder.textView = view.findViewById(R.id.event_text);
			viewHolder.shareButton = view.findViewById(R.id.event_share);
			view.setTag(viewHolder);
		}

		final Event event = getCurrentEvents().get(position);
		viewHolder.iconView.setImageResource(event.getCategory().getImageRes());
		viewHolder.categoryTextView.setText(event.getCategory().getDisplayNameRes());
		viewHolder.textView.setText(Html.fromHtml(event.getText()));
		viewHolder.textView.setMovementMethod(LinkMovementMethod.getInstance());
		viewHolder.shareButton.setOnClickListener(v ->
		{
			Intent sendIntent = new Intent();
			sendIntent.setAction(Intent.ACTION_SEND);
			sendIntent.putExtra(Intent.EXTRA_TEXT, event.getText());
			sendIntent.setType("text/plain");

			Intent shareIntent = Intent.createChooser(sendIntent, null);
			v.getContext().startActivity(shareIntent);
		});

		return view;
	}

	private static class ViewHolder
	{
		ImageView iconView;
		TextView categoryTextView;
		TextView textView;
		ImageButton shareButton;
	}
}

File: android/app/src/main/java/com/tisza/esemenynaptar/EventListAdapter.kt
package com.tisza.esemenynaptar

import android.content.*
import android.os.*
import android.text.*
import android.text.method.*
import android.view.*
import android.widget.*
import androidx.lifecycle.*
import com.tisza.esemenynaptar.database.*

class EventListAdapter(private val context: Context, private val eventRes: Int, private val noEventRes: Int) : BaseAdapter() {
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val noEventsView = inflater.inflate(noEventRes, null, false)
    private val observer = Observer<List<Event>> { notifyDataSetChanged() }

    var events: LiveData<List<Event>>? = null
        set(value) {
            field?.removeObserver(observer)
            field = value
            field?.observeForever(observer)
        }

    var onEventClickedListener: ((Event) -> Unit)? = null

    private val currentEvents: List<Event>
        get() = events?.value ?: emptyList()

    override fun getCount(): Int {
        return currentEvents.size.coerceAtLeast(1)
    }

    override fun getItem(position: Int): Any? {
        return null
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        if (currentEvents.isEmpty())
            return noEventsView

        val view: View
        val viewHolder: ViewHolder
        if (convertView != null && convertView !== noEventsView) {
            view = convertView
            viewHolder = convertView.tag as ViewHolder
        } else {
            view = inflater.inflate(eventRes, parent, false)
            viewHolder = ViewHolder()
            viewHolder.categoryTextView = view.findViewById(R.id.event_category_text)
            viewHolder.iconView = view.findViewById(R.id.event_icon)
            viewHolder.dateTextView = view.findViewById(R.id.event_date)
            viewHolder.textView = view.findViewById(R.id.event_text)
            viewHolder.shareButton = view.findViewById(R.id.event_share)
            viewHolder.likeButton = view.findViewById(R.id.event_like)
            view.tag = viewHolder
        }
        val event = currentEvents[position]

        view.setOnClickListener {
            onEventClickedListener?.invoke(event)
        }
        viewHolder.iconView.setImageResource(event.category.imageRes)
        viewHolder.categoryTextView.setText(event.category.displayNameRes)
        viewHolder.dateTextView.text = dateFormat.format(event.calendar.time)
        viewHolder.textView.text = Html.fromHtml(event.text)
        viewHolder.textView.movementMethod = if (onEventClickedListener == null) LinkMovementMethod.getInstance() else null
        viewHolder.shareButton.setOnClickListener { v: View ->
            val sendIntent = Intent()
            sendIntent.action = Intent.ACTION_SEND
            sendIntent.putExtra(Intent.EXTRA_TEXT, event.text)
            sendIntent.type = "text/plain"
            val shareIntent = Intent.createChooser(sendIntent, null)
            v.context.startActivity(shareIntent)
        }
        viewHolder.likeButton.alpha = if (event.isLiked) 1.0f else 0.3f
        viewHolder.likeButton.setOnClickListener {
            object : AsyncTask<Unit, Unit, Unit>() {
                override fun doInBackground(vararg p0: Unit) {
                    event.isLiked = !event.isLiked
                    eventDatabase.eventDao().updateEvent(event)
                }
            }.execute()
        }
        return view
    }

    private class ViewHolder {
        lateinit var iconView: ImageView
        lateinit var categoryTextView: TextView
        lateinit var dateTextView: TextView
        lateinit var textView: TextView
        lateinit var likeButton: ImageButton
        lateinit var shareButton: ImageButton
    }
}
